README.md

```diff
@@ -4,57 +4,120 @@ This is a tiny little shim around Cosmos SDK binaries that use the upgrade
+
+Note that chains that wish to support upgrades may package up a genesis upgrade manager tar file with this info, just as they
+prepare the genesis binary tar file. In fact, they may offer a tar file will all upgrades up to current point for easy download
+for those who wish to sync a fullnode from start.
+
+The `DAEMON` specific code, like the tendermint config, the application db, syncing blocks, etc is done as normal.
+The same eg. `GAIA_HOME` directives and command-line flags work, just the binary name is different.
 
 ## Upgradeable Binary Specification
 
-This upgrade manager works with any binary that follows the following specification:
+In the basic version, the upgrade_manager will read the stdout log messages
+to determine when an upgrade is needed. We are considering more complex solutions
+via signaling of some sort, but starting with the simple design:
 
-* when an upgrade is needed the binary will emit a message that matches this
+* when an upgrade is needed the binary will print a line that matches this
 regular expression: `UPGRADE "(.*)" NEEDED at height (\d+):(.*)`.
 * the second match in the above regular expression can be a JSON object with
 a `binaries` key as described above
 
-Binaries will be started as sub-processes and killed when an upgrade is needed
+The name (first regexp) will be used to select the new binary to run. If it is present,
+the current subprocess will be killed, `current` will be upgraded to the new directory, 
+and the new binary will be launched.
+
+**Question** should we just kill the upgrade manager after it does the updates?
+so it gets a clean restart and just runs the new binary (under `current`).
+it should be safe to restart (as a service).
+
+## Auto-Download
+
+Generally, the system requires that the administrator place all relevant binaries
+on the disk before the upgrade happens. However, for people who don't need such
+control and want an easier setup (maybe they are syncing a non-validating fullnode
+and want to  do little maintenance), there is another option.
+
+If you set `DAEMON_DOWNLOAD=on` then when an upgrade is triggered and no local binary
+can be found, the upgrade_manager will attempt to download and install the binary itself.
+The plan stored in the upgrade module has an info field for arbitrary json.
+This info is expected to be outputed on the halt log message. There are two
+valid format to specify a download in such a message:
+
+1. Store an os/architecture -> binary URI map in the upgrade plan info field
+as JSON (or YAML??) under the `"binaries"` key, eg:
+```json
+{
+  "binaries": {
+    "linux/amd64":"https://example.com/gaiad?checksum=sha256:b7d96c89d09d9e204f5fedc4d5d55b21"
+  }
+}
+```
+2. Store a link to a file that contains all information in the above format (eg. if you want
+to specify lots of binaries, changelog info, etc without filling up the blockchain).
+```json
+{
+  "link":"https://example.com/testnet-1001-info.json"
+}
+```
```

Comment:
Let's not make this link embedded in JSON. I would say either a URL or JSON, not JSON with a URL

Reply:
The example link URI should also include a checksum

Reply:
It is kind of annoying to have a string that may or may not be json.
Just for parsers.

But I will update there.